src/lib/storage/variable_string_dictionary/variable_string_dictionary_encoder.hpp

```diff
@@ -0,0 +1,122 @@
+#include "storage/value_segment.hpp"
+#include "storage/variable_string_dictionary_segment.hpp"
+#include "storage/vector_compression/base_compressed_vector.hpp"
+#include "storage/vector_compression/vector_compression.hpp"
+#include "types.hpp"
+#include "utils/enum_constant.hpp"
+
+namespace hyrise {
+
+/**
+ * @brief Encodes a segment using variable string dictionary encoding and compresses its attribute vector using vector compression.
+ *
+ * The algorithm first creates an attribute vector of standard size (uint32_t) and then compresses it
+ * using fixed-width integer encoding.
+ */
+class VariableStringDictionaryEncoder : public SegmentEncoder<VariableStringDictionaryEncoder> {
+ public:
+  static constexpr auto _encoding_type = enum_c<EncodingType, EncodingType::VariableStringDictionary>;
+  static constexpr auto _uses_vector_compression = true;  // see base_segment_encoder.hpp for details
+
+  std::shared_ptr<AbstractEncodedSegment> _on_encode(const AnySegmentIterable<pmr_string> segment_iterable,
+                                                     const PolymorphicAllocator<pmr_string>& allocator) {
+    // Vectors to gather the input segment's data. This data is used in a later step to
+    // construct the actual dictionary and attribute vector.
+    auto dense_values = std::vector<pmr_string>();  // Contains the actual values (no NULLs).
+    auto null_values = std::vector<bool>();         // bitmap to mark NULL values
+    // Maps string to ChunkOffsets for faster write of vector that maps ChunkOffsets to ValueID.
+    auto string_to_chunk_offsets = std::unordered_map<pmr_string, std::vector<ChunkOffset>>();
+    auto segment_size = uint32_t{0};
+
+    // Iterate over segment, save all values and save to values the chunk_offset.
+    segment_iterable.with_iterators([&](auto segment_it, const auto segment_end) {
+      segment_size = std::distance(segment_it, segment_end);
+      dense_values.reserve(segment_size);  // Potentially overallocate for segments with NULLs.
+      null_values.resize(segment_size);    // Resized to size of segment.
+
+      for (auto current_position = size_t{0}; segment_it != segment_end; ++segment_it, ++current_position) {
+        const auto segment_item = *segment_it;
+        if (!segment_item.is_null()) {
+          const auto& segment_value = segment_item.value();
+          dense_values.push_back(segment_value);
+          string_to_chunk_offsets[segment_value].push_back(ChunkOffset(current_position));
+        } else {
+          null_values[current_position] = true;
+        }
+      }
+    });
+
+    // Eliminate duplicate strings.
+    std::sort(dense_values.begin(), dense_values.end());
+    dense_values.erase(std::unique(dense_values.begin(), dense_values.end()), dense_values.cend());
+    dense_values.shrink_to_fit();
+
+    // Compute total compressed data size.
+    auto total_size = uint32_t{0};
+    for (const auto& value : dense_values) {
+      total_size += value.size() + 1;
+    }
```

Comment:
Untested, but could also written like this:
```suggestion
    auto total_size = std::accumulate(dense_values.begin(), dense_values.end(), uint32_t{0}, [](uint32_t acc, pmr_string& value) { return acc + value.size() + 1; });
```